java/code/src/com/suse/manager/maintenance/MaintenanceManager.java

```diff
@@ -96,6 +106,101 @@ public static MaintenanceManager instance() {
         return instance;
     }
 
+    /**
+     * Given the systems and action type, return upcoming maintenance windows.
+     *
+     * The upper limit of returned maintenance windows is currently hardcoded to 10.
+     *
+     * If given action type is not maintenance-mode-only, return an empty optional.
+     * If given systems do not have any maint. <b>schedules</b> assigned, return an empty optional.
+     * If given systems have different maint. schedules assigned, throw an exception.
+     * Otherwise return list of maintenance windows (the list can be empty, if the schedule does not contain
+     * any upcoming maintenance windows).
+     *
+     * @param actionType the action type
+     * @param systemIds the system ids
+     * @return the optional upcoming maintenance windows
+     * @throws IllegalStateException if two or more systems have different maint. schedules assigned
+     */
+    public Optional<List<Triple<String, String, Long>>> calculateUpcomingMaintenanceWindows(ActionType actionType,
+            Set<Long> systemIds) throws IllegalStateException {
+        // non-maintenance mode only don't use maint. windows
+        if (!actionType.isMaintenancemodeOnly()) {
+            return empty();
+        }
+
+        Set<MaintenanceSchedule> schedules = MaintenanceManager.instance().listSchedulesOfSystems(systemIds);
+        // if there are no schedules, we don't populate any windows (and display an empty list instead)
+        if (schedules.isEmpty()) {
+            return empty();
+        }
+
+        // there are multiple schedules, we warn the user about it
+        if (schedules.size() > 1) {
+            throw new IllegalStateException("Multiple schedules");
+        }
+
+        PeriodList periods = schedules.iterator().next().getCalendarOpt()
+                .flatMap(c -> parseCalendar(c))
+                .map(c -> calculateUpcomingPeriodList(c, empty(), new Date(), 10))
```

Comment:
`empty()` is not correct here, right? You need to check if schedules has a multi calendar and if yes, pass the schedule name to it.

Reply:
Good catch! This overflowed here from the prototype somehow :)